Lattice/State Refactoring & generic states

gitsimon opened the pull request:
# Refactoring of Lattice/State

- remove internal representation
- sub-classing Lattice provides you with a implicit Top/Bottom element as before, subclassing BaseLattice leaves you with complete freedom to implement a valid lattice yourself.

Add generic lattices for

- store: variable -> other_lattice
- stack of lattice elements

# Adapt examples

- Dummy Example (playground) adapted to use generic lattices
- LiveDead adapted to use generic lattices (StoreLattice)
- Usage analysis implemented, **working**!

# CFG generator from python source code

- Partial implementation of CFG generator without handling complex cases like classes, multiple assignment targets etc.
- the `ControlFlowGraph` class is untouched, for the generation necessary _loose control flow graphs_ are represented by a separate class inside `cfg_generator.py`

# Unit test setup

- generic test cases that makes it easy to add test cases that auto scan all sub directories of specified directory for python source files and runs a desired analysis on them 
  - generate CFG
  - run analysis
  - check actual analysis result against expected results annotated by comments in the source of the form `# RESULT: ...`

gitsimon committed:
UsedLattice

gitsimon committed:
add print expression

gitsimon committed:
used-analysis states

gitsimon committed:
Merge remote-tracking branch 'upstream/master'

gitsimon committed:
refactor lattice

remove internal representation completely

gitsimon committed:
Merge remote-tracking branch 'upstream/master' into lattice-refactoring

# Conflicts:
#	abstract_domains/liveness.py

gitsimon committed:
fix replace method

gitsimon committed:
add generators for convenient iteration to cfg

unused so far

gitsimon committed:
adapt forward/backward interpreter

distinguish between bottom in state and 'not yet set' since refactored lattice (and derived states) do not provide this distinction

gitsimon committed:
add generic lattices, fix playground

gitsimon committed:
reorganize files

gitsimon committed:
adapt usage analysis (incomplete)

gitsimon committed:
remove explicit print expression again

print can now be implemented via UserDefinedCallSemantics

gitsimon committed:
add enter_if/exit_if hooks

and add Edge kinds for this

gitsimon committed:
adapt and fix used analysis

gitsimon committed:
renaming

follow python convention with capital enum constants

gitsimon committed:
separate usage semantics

gitsimon committed:
fix bugs and inspection warnings

gitsimon committed:
cfg generator

caterinaurban commented on abstract_domains/state.py:
```diff
@@ -1,53 +1,28 @@
-    def internal(self):
-        return self._internal
+        self._result = set()  # set of expressions representing the result of the previously analyze statement
+        super().__init__()
```
Why is the super call last?

gitsimon replied:
fixed

caterinaurban commented on abstract_domains/state.py:
```diff
@@ -1,53 +1,28 @@
-        
-        :return: string representing the analysis state
-        """
+    def __repr__(self):
```
Can we put the docstring that __str__ had here?

gitsimon replied:
not the docstring from `__str__`, but another meaningful description added

caterinaurban commented on abstract_domains/state.py:
```diff
@@ -1,53 +1,28 @@
-            self._result = result
-
-    def __init__(self, kind: Lattice.Kind = Lattice.Kind.Default):
+class State(Lattice, ABC):
```
I think this should extend the Lattice with Top and Bottom. But we should talk about this before making more changes.

gitsimon replied:
Lattice is now the default WITH Top and Bottom element. If somebody do not want the this, they can still subclass BaseLattice.

caterinaurban commented on abstract_domains/liveness/livedead.py:
```diff
@@ -1,46 +1,49 @@
         return "{}".format(variables)
 
+    def default(self):
+        return self.bottom()
```
The implementation here should be the implementation that bottom() has at the moment.

gitsimon replied:
fixed

caterinaurban commented on abstract_domains/liveness/livedead.py:
```diff
@@ -1,46 +1,49 @@
+
+    def bottom(self):
+        self._variables = {variable: LiveDead.Liveness.Dead for variable in self.variables}
+        return self
```
Related to the comment on State, this should be the special Bottom element of the Lattice.  We should talk about this.

gitsimon replied:
fixed

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -1,93 +1,74 @@
-        return self.internal.kind
+        :return: human-readable string
+        """
+        return repr(self)
```
We can get rid of the __str__ method if it just calls the __repr__ method since I think it does that by default.

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -188,10 +169,11 @@ def widening(self, other: 'Lattice'):
         """
-        self._internal = other._internal
+        self.__dict__.update(other.__dict__)
+        assert self == other
```
Is this why you added the __eq__ method implementation?

gitsimon replied:
yes. the assert check would actually fail without overwriting __eq__. It is anyway a good idea to overwrite __eq__ in this case. The default implementation uses object addresses to check for equality, but a lattice element is usually more perceived as a "value" which should be comparable like e.g. an integer.

caterinaurban commented on abstract_domains/usage/used.py:
```diff
@@ -0,0 +1,115 @@
+        self._used = U
+        return self
+
+    def is_bottom(self) -> 'bool':
```
No need for the ' around bool

gitsimon replied:
fixed

caterinaurban commented on abstract_domains/usage/used.py:
```diff
@@ -0,0 +1,115 @@
+    def is_bottom(self) -> 'bool':
+        return self.used == N
+
+    def is_top(self) -> 'bool':
```
No need for the ' around bool

caterinaurban commented on abstract_domains/usage/used.py:
```diff
@@ -0,0 +1,115 @@
+    U = 3
+    UU = 2
+    UN = 1
+    N = 0
```
Can you add comments near the flags to clarify what they represent?

gitsimon replied:
done

caterinaurban commented on abstract_domains/usage/used.py:
```diff
@@ -0,0 +1,115 @@
+        (U, UU): U,
+        (U, UN): UN,
+        (U, U): U,
+    }
```
What kind of code construct are these _DESCEND and _COMBINE? I have never seen such a thing :D

gitsimon replied:
dictionaries

caterinaurban commented on abstract_domains/usage/used.py:
```diff
@@ -0,0 +1,115 @@
+        if self.used == other.used or self.used == N:
+            return True
+        elif (self.used == UU and other.used == UN) or (
+                        self.used == UN and other.used == UU):
```
(nitpicking, feel free to ignore)
Is the formatting meant to be so ugly ?

gitsimon replied:
auto-formatter is not always right :)

caterinaurban commented on core/cfg.py:
```diff
@@ -165,19 +161,26 @@ def __str__(self):
         :param in_node: entry node of the control flow graph
         :param out_node: exit node of the control flow graph
-        :param edges: edges of the control flow graph
+        :param edges: optional set of edges of the control flow graph
```
Why are these optional?

gitsimon replied:
changed back to your original version

caterinaurban commented on core/cfg.py:
```diff
@@ -165,19 +161,26 @@ def __str__(self):
-        :param edges: edges of the control flow graph
+        :param edges: optional set of edges of the control flow graph
+        :param loose_in_edges: optional set of loose edges that have no start yet and end inside this CFG
+        :param loose_out_edges: optional set of loose edges that start inside this CFG and have no end yet
```
What are these things? Did you add them just for the CFG generation? Can we have the stuff related to the CFG generation in the CFG generator instead?

gitsimon replied:
extracted out in frontend.cfg_generator (use the original, complete CFG by aggregation)

caterinaurban commented on engine/backward.py:
```diff
@@ -36,7 +36,7 @@ def analyze(self, initial: State) -> AnalysisResult:
                 previous = deepcopy(self.result.get_node_result(current)[-1])
             else:
-                previous = deepcopy(initial).bottom()
+                previous = None
```
I am still confused about this. We should talk about it ;)

caterinaurban reviewed (commented):
A lot of comments but mostly about the same thing. We should discuss about it tomorrow :)

gitsimon committed:
add unittest and fixes

gitsimon committed:
fix engine when unvisited predecessor/successors

in more complex control flow graphs it could happen that we visit a node with unvisited predecessors/successors. In this case we now put that node to queue again and  continue with next node.

gitsimon committed:
engine: remove unnecessary None checks

gitsimon committed:
separate loose CFG

since loose CFGs are only used during generation, they are now separated from ControlFlowGraph. The LooseControlFlowGraph extends the functionality of the complete ControlFlowGraph via aggregation.

gitsimon committed:
unittests: check actual against expected results provided in comments

gitsimon committed:
extract generic testcase functions

gitsimon committed:
class based (generic) unittests

extract common testcase functionality in super classes
it is now easy to create new test cases for specific analysis, giving you the freedom of running pre/post analysis but providing the uninteresting boilerplate code.

gitsimon committed:
rename used elements and comment

to U, S, O, N

gitsimon committed:
refactor lattices

gitsimon committed:
adapt liveness to use generic StoreLattice

gitsimon committed:
fixes from review, fix inspection warnings

gitsimon committed:
unittests: fix actual result search

consistently search for first actual result AFTER the expected result comment
